File: 1970-last-day-where-you-can-still-cross/1970-last-day-where-you-can-still-cross.java
class Solution {
    public int latestDayToCross(int row, int col, int[][] cells) {
        int[][] grid = new int[row][col];
        int lo = 0, hi = row*col;
        while(lo < hi){
            int mid = (lo + hi + 1) >> 1;
            for (int i = 0; i < mid; i++){
                grid[cells[i][0]-1][cells[i][1]-1] = mid; // mark visited
            }
            boolean ok = false;
            for (int i = 0; i < col && !ok; i++){
                ok = dfs(0, i, grid, mid);
            }
            if (ok)
                lo = mid;
            else
                hi = mid - 1;
        }
        return lo;
    }

    private boolean dfs(int i, int j, int[][] grid, int visited){
        if (i < 0 || j < 0 || i == grid.length || j == grid[0].length || grid[i][j] == visited)
            return false;
        grid[i][j] = visited;
        return i == grid.length - 1
            || dfs(i + 1, j, grid, visited)
            || dfs(i - 1, j, grid, visited)
            || dfs(i, j + 1, grid, visited)
            || dfs(i, j - 1, grid, visited);
    }
}

File: 1970-last-day-where-you-can-still-cross/1970-last-day-where-you-can-still-cross.cpp
class Solution {
public:
    bool isPathExist(int x,int y,vector<vector<bool>> &vis,vector<vector<int>> &grid){
        vis[x][y]=true;
        if(x==grid.size()-1) return true;
        int delR[] = {-1,1,0,0};
        int delC[] = {0,0,-1,1};
        for(int k=0;k<4;k++){
            int nr = delR[k]+x;
            int nc = delC[k]+y;
            if(nr>=0&&nc>=0&&nr<grid.size()&&nc<grid[0].size()
               &&!vis[nr][nc]&&grid[nr][nc]==0){
                if(isPathExist(nr,nc,vis,grid)==true) return true;
            }
        }
        return false;
    }
    bool isValid(int mid,int row,int col,vector<vector<int>>& cells){
        vector<vector<int>> grid(row,vector<int>(col,0));
        for(int i=0;i<mid;i++) grid[cells[i][0]-1][cells[i][1]-1]=1;
        vector<vector<bool>> vis(row,vector<bool>(col,false));
        for(int i=0;i<col;i++){
            if(grid[0][i]==0 && !vis[0][i] && isPathExist(0,i,vis,grid)==true) return true;
        }
        return false;
    }
    int latestDayToCross(int row, int col, vector<vector<int>>& cells) {
        int lo = 1;
        int hi = cells.size();
        int ans = 0;
        while(lo<=hi){
            int mid = (lo+hi)/2;
            if(isValid(mid,row,col,cells)==true){
                ans = lo;
                lo = mid+1;
            }else{
                hi=mid-1;
            }
        }
        if(isValid(hi,row,col,cells)==true) return hi;
        else return ans;
    }
};